### PR TITLE
Cluster-state based Security role mapper

### PR DESCRIPTION
**Pull request by albertzaharovits**
This implements a new `UserRoleMapper` that sources the role mapping rules from the cluster state.
The role mapping rules are stored under a new custom cluster state that **is** persisted (both disk and snapshots).
The role mapper refreshes realm caches when role mapping changes are published.
The role mapper is disabled by default, and it can only be enabled from code, by other plugins.
When enabled, the cluster state role mappings rules, if any, are additive to the rules from
the index-based native role mapping store and the file-based DN one.

### CODE REVIEW

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/support/mapper/CompositeRoleMapperTests.java [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+public class CompositeRoleMapperTests extends ESTestCase {
+
+    public void testClearRealmCachePropagates() {
+        UserRoleMapper userRoleMapper1 = mock(UserRoleMapper.class);
+        UserRoleMapper userRoleMapper2 = mock(UserRoleMapper.class);
+        CompositeRoleMapper compositeRoleMapper = new CompositeRoleMapper(userRoleMapper1, userRoleMapper2);
+        CachingRealm realm = mock(CachingRealm.class);
+        compositeRoleMapper.clearRealmCacheOnChange(realm);
+        verify(userRoleMapper1, times(1)).clearRealmCacheOnChange(eq(realm));
+        verify(userRoleMapper1, times(1)).clearRealmCacheOnChange(eq(realm));
```
Comment (gwbrown) [inline x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/support/mapper/CompositeRoleMapperTests.java:38]: I think you meant:
```suggestion
        verify(userRoleMapper1, times(1)).clearRealmCacheOnChange(eq(realm));
        verify(userRoleMapper2, times(1)).clearRealmCacheOnChange(eq(realm));
```

?

**Review by AthenaEryma — APPROVED** [review]
Thanks for your patience on this review. Nice work, I left a couple comments but nothing major - no need for another round of review once they're addressed to your satisfaction.

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/support/mapper/CompositeRoleMapperTests.java (hunk 1 above)
Comment (albertzaharovits) [reply x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/support/mapper/CompositeRoleMapperTests.java:38]: Good catch! Thanks!

File: x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/xpack/security/authc/jwt/JwtRoleMappingsIntegTests.java [hunk 2]
```diff
@@ -0,0 +1,483 @@
...
+public final class JwtRoleMappingsIntegTests extends SecurityIntegTestCase {
+
+    private final String jwt0SharedSecret = "jwt0_shared_secret";
+    private final String jwt1SharedSecret = "jwt1_shared_secret";
+    private final String jwtHmacKey = "test-HMAC/secret passphrase-value";
+    private static boolean anonymousRole;
+
+    @BeforeClass
+    public static void beforeTests() {
+        anonymousRole = randomBoolean();
```
[the remark this answers is not in the document]
Comment (albertzaharovits) [reply x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/xpack/security/authc/jwt/JwtRoleMappingsIntegTests.java:62]: The `xpack.security.authc.anonymous.roles` is a static setting, so it's cumbersome to vary from test to test. It's also not super relevant for the scope of the tests.

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/RoleMappingMetadataTests.java [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+import org.elasticsearch.xpack.core.security.authc.support.mapper.ExpressionRoleMapping;
+import org.elasticsearch.xpack.core.security.authz.RoleMappingMetadata;
+
+import java.io.IOException;
+import java.util.Set;
+
+import static org.elasticsearch.xpack.security.authc.support.mapper.ExpressionRoleMappingTests.randomRoleMapping;
+import static org.hamcrest.Matchers.is;
+
+public class RoleMappingMetadataTests extends ESTestCase {
```
[the remark this answers is not in the document]
Comment (albertzaharovits) [reply x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/RoleMappingMetadataTests.java]: Good point, pushed: https://github.com/elastic/elasticsearch/pull/107410/commits/4135aeb792db8e517d77bb35128e2f82097d84ae

**Pull request merged**